[EXPERIMENTAL] Use DomainMatrix internally for matrices of integers/rationals

oscarbenjamin opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Uses the DomainMatrix class added in #18844

#### Brief description of what is fixed or changed

Makes a Matrix/ImmutableMatrix of all Integers/Rationals return a subclass of DenseMatrix that uses DomainMatrix internally.

#### Other comments

This is an experimental PR testing out what it could be like to use DomainMatrix internally for matrix calculations. I do not intend to implement something like this but rather to make a new matrix class called e.g. `Mat` that would be immutable, would use `DomainMatrix` internally and would be completely separate from the existing matrix classes.

This PR can be used to test out the possible benefits of using DomainMatrix and develop its interface. It also highlights a number of issues that would also need to be fixed as part of this work.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin commented:
I've added an inverse method for `DomainMatrix` and made sure that `Matrix.inv()` calls through to it (when the matrix is all rational/integer). This is much faster than the current default implementation of inverse.

Script:
```python
from sympy import *
import time

for n in 1, 2, 5, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20:
    M = randMatrix(n)
    start = time.time()
    Minv = M.inv()
    delta = time.time() - start
    assert M * Minv == eye(n)
    print('n = %2d, t = %.3g secs' % (n, delta))
```

Timings with the PR:
```console
$ python t.py 
n =  1, t = 3.81e-05 secs
n =  2, t = 5.41e-05 secs
n =  5, t = 0.000262 secs
n = 10, t = 0.00226 secs
n = 11, t = 0.00288 secs
n = 12, t = 0.00208 secs
n = 13, t = 0.00262 secs
n = 14, t = 0.00448 secs
n = 15, t = 0.00525 secs
n = 16, t = 0.00536 secs
n = 17, t = 0.00688 secs
n = 18, t = 0.00811 secs
n = 19, t = 0.0506 secs
n = 20, t = 0.0122 secs
```
Timings on master:
```console
$ python t.py 
n =  1, t = 0.00109 secs
n =  2, t = 0.00159 secs
n =  5, t = 0.0071 secs
n = 10, t = 0.0486 secs
n = 11, t = 0.0851 secs
n = 12, t = 0.122 secs
n = 13, t = 0.454 secs
n = 14, t = 0.716 secs
n = 15, t = 2.47 secs
n = 16, t = 9.82 secs
n = 17, t = 19.7 secs
n = 18, t = 82.1 secs
^C
```

oscarbenjamin committed:
wip

sylee957 commented:
It is not easy to extend things beyond cholesky decomposition for this because it needs square roots.
Although there is AlgebraicElement for this, I’m not sure it can be used for zero testing

oscarbenjamin commented:
Probably Cholesky just needs to work over EX. The LDL decomposition can avoid square roots.

I think zero testing works fine over algebraic fields:
```julia
In [11]: K = AlgebraicField(QQ, sqrt(2))                                                                                                       

In [12]: e = K.convert(1+sqrt(2))                                                                                                              

In [13]: e                                                                                                                                     
Out[13]: ANP([mpq(1,1), mpq(1,1)], [mpq(1,1), mpq(0,1), mpq(-2,1)], QQ)

In [14]: (e - 1)**2 - 2 == K.zero                                                                                                              
Out[14]: True
```
The problem though is that in the Cholesky algorithm each row might make a new square root requiring a new extension field. It seems that AlgebraicField gets slower as the degree of the extension increases:
```julia
In [68]: M = Matrix([[3, 1, 1], [1, 4, 1], [1, 1, 5]])                                                                                         

In [69]: M.cholesky()                                                                                                                          
Out[69]: 
Matrix([
[  sqrt(3),             0,             0],
[sqrt(3)/3,    sqrt(33)/3,             0],
[sqrt(3)/3, 2*sqrt(33)/33, 5*sqrt(22)/11]])

In [70]: K, (r2, r3, r11) = construct_domain([sqrt(2), sqrt(3), sqrt(11)], extension=True)                                                     

In [71]: _ = list(map(K.convert, M.cholesky()._mat))   # <--------- slow

In [72]: [K.to_sympy(x) for x in _]
Out[72]: 
⎡          √3  √33     √3  2⋅√33  5⋅√22⎤
⎢√3, 0, 0, ──, ───, 0, ──, ─────, ─────⎥
⎣          3    3      3     33     11 ⎦
```
Also we can only have an `AlgebraicField` over QQ so this wouldn't work if the matrix has any symbols.

It might be that Cholesky is just not as suitable for symbolic computation as it is for numerics. It looks like at least LDL can work in any normal field.

Probably `DomainMatrix` is more suitable for immutable matrices. I just patched it into mutable Matrix here because that's more widely used and I want to see the results of the whole existing test suite and benchmarks for matrices.

sylee957 commented:
What are the key difficulties of implementing the mutable version than immutable version?

oscarbenjamin commented:
> What are the key difficulties of implementing the mutable version than immutable version?

The main problem is if setting an item means that the domain changes. In this PR I've changed many routines to use lists rather than matrices as a workaround. Here if you do
```
M = zeros(n)
M[0, 0] = x
```
then zeros gives a matrix that internally has a DomainMatrix whose domain is QQ. If `x` is a symbol or `sqrt(2)` or anything else that is not Rational then it can not be represented in QQ so in this PR that raises. I don't intend for that to be what happens with mutable matrices: I've just implemented it that way in this PR as a workaround and then rewritten any routines that fail as a result.

We could implement a mutable matrix that detects when something can't be converted to the domain and then calls unify and recreates a new internal DomainMatrix converting all of the elements from one domain to another but there are problems with that:

1. It could be very expensive having to reconvert all elements of the matrix repeatedly in `__setitem__`.
2. This would prevent some optimisations such as deciding at the start of a routine how to proceed from knowing the domain.

Really I think a better approach is to make a standalone efficient immutable matrix and then a calculation (e.g. rref) with a mutable matrix can convert to the immutable matrix once at the start and then use the efficient calculations internally. This means that we make an immutable matrix and then use that to implement mutable matrices. I think that is better than the current approach which is implements immutable matrices by calling `as_mutable` and `_new` at the start and end of a routine and then does all the internal calculations with mutable matrices.

sylee957 commented:
I thought that Domain can rather be a immutable parameter for mutable matrix and refuse elements that can’t be coerced.

oscarbenjamin commented:
> I thought that Domain can rather be a immutable parameter for mutable matrix and refuse elements that can’t be coerced.

That's what happens in this PR. I don't think many users of mutable matrix would be happy with that though. The most common use of mutability within sympy is to build a matrix starting with zeros and then using assignments. What should the domain be if you call zeros?

sylee957 commented:
Although the sympy community may encourage to switch to immutable matrices.
I'm afraid that many people are ultimately going to get confused about how they find anything useful for immutable only API.

For how numpy deals with the issue, I see it tries to round to integers if given float elements and refuses complex elements, so they may have the best idea about how to have a mutable array of fixed data types.
```
>>> import numpy as np
>>> a = np.zeros((3, 3), dtype=np.int32)
>>> a[0, 0] = 1j
TypeError: can't convert complex to int
```

oscarbenjamin commented:
The problem is that numpy's data-types are relatively understandable e.g. most users who are aware of the dtypes probably know that they can have int or float and that they need to specify which they want. Here specifying the "dtype" as `QQ_I[x, y]` is probably too complicated for most of our users.

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin commented:
This PR also makes matrix multiplication roughly 10x faster for integer matrices:
```python
from sympy import *
import time

for n in [1, 10, 100, 200]:
    start = time.time()
    M1 = randMatrix(n)
    M2 = randMatrix(n)
    M1*M2
    delta = time.time() - start
    print('n = %2d, t = %.3g secs' % (n, delta))
```
On master:
```
n =  1, t = 0.000363 secs
n = 10, t = 0.0155 secs
n = 100, t = 7.86 secs
n = 200, t = 61.9 secs
```
With the PR:
```
n =  1, t = 0.000501 secs
n = 10, t = 0.00294 secs
n = 100, t = 0.625 secs
n = 200, t = 4.58 secs
```

oscarbenjamin commented:
The benchmarks are showing a lot of slowdowns though:
https://travis-ci.org/github/sympy/sympy/jobs/704977797#L1223

oscarbenjamin commented:
Looking at the benchmarks the cases that have slowed down are those that are very fast. I think there's a little bit of overhead in checking whether the elements of the matrix are rational.

oscarbenjamin committed:
wip

sylee957 commented:
I’ve thought that a more incremental way of making changes in development is that using domainmatrix internally from the matrix classes we originally have than have a huge PR introducing complete new matrices.

If the overhead only applies for small matrix, I think that the conversion can be strategically done with respect to the matrix size.

oscarbenjamin commented:
> I’ve thought that a more incremental way of making changes in development is that using domainmatrix internally from the matrix classes we originally have than have a huge PR introducing complete new matrices.

Maybe. We need to build up `DomainMatrix` into something that is usable though which is the purpose of this PR. One possible approach is to have all mutable matrices use `DomainMatrix` internally and then speed up mutable calculations by converting to immutable.

It's possible that there could be a class in between `DomainMatrix` and `ImmutableMatrix` so that `DomainMatrix` is just for calculations within some domain but the next level up handles conversions and unification etc like `Poly` does. Then that could be used directly by users or internally by ImmutableMatrix.

> If the overhead only applies for small matrix, I think that the conversion can be strategically done with respect to the matrix size.

I think that the overheads are just to do with the way I have patched the constructor because this ends up calling the normal matrix constructor and then afterwards checking if everything is rational and then calling the domainmatrix constructor. It adds a tiny overhead to trivial (`<1ms`) calculations. I don't think that overhead would be there if this was implemented in a different way and I don't intend to implement it this way in general.

oscarbenjamin committed:
wip - use domainmatrix for rank

oscarbenjamin commented:
The last commit uses `DomainMatrix.rref` for `rank` which massively speeds up calculating the rank of integer matrices:
```python
from sympy import *
import time

for n in [1, 2, 5, 10, 15, 20, 21, 22, 23, 24, 25]:
    start = time.time()
    M1 = randMatrix(n)
    M1.rank()
    delta = time.time() - start
    print('n = %2d, t = %.3g secs' % (n, delta))
```
On master:
```
n =  1, t = 0.00035 secs
n =  2, t = 0.000865 secs
n =  5, t = 0.00504 secs
n = 10, t = 0.0162 secs
n = 15, t = 0.0476 secs
n = 20, t = 3.19 secs
n = 21, t = 11.9 secs
n = 22, t = 35.8 secs
^C
```
With the PR:
```
n =  1, t = 0.000313 secs
n =  2, t = 0.000304 secs
n =  5, t = 0.000621 secs
n = 10, t = 0.00154 secs
n = 15, t = 0.00349 secs
n = 20, t = 0.00747 secs
n = 21, t = 0.00826 secs
n = 22, t = 0.0108 secs
n = 23, t = 0.0121 secs
n = 24, t = 0.0128 secs
n = 25, t = 0.0135 secs
```

oscarbenjamin commented:
> On master:
> 
> ```
> n =  1, t = 0.00035 secs
> n =  2, t = 0.000865 secs
> n =  5, t = 0.00504 secs
> n = 10, t = 0.0162 secs
> n = 15, t = 0.0476 secs
> n = 20, t = 3.19 secs
> n = 21, t = 11.9 secs
> n = 22, t = 35.8 secs
> ^C
> ```

I don't understand why so many calculations with Matrix scale so badly. There shouldn't be any issue with blowup as these are matrices of integers. Also finding a pivot should be trivial for integers.

Maybe it's because of using division-free reduction which leads to integer blowup. Not sure but the issue but it also applies to computing the inverse.

> With the PR:
> 
> ```
> n =  1, t = 0.000313 secs
> n =  2, t = 0.000304 secs
> n =  5, t = 0.000621 secs
> n = 10, t = 0.00154 secs
> n = 15, t = 0.00349 secs
> n = 20, t = 0.00747 secs
> n = 21, t = 0.00826 secs
> n = 22, t = 0.0108 secs
> n = 23, t = 0.0121 secs
> n = 24, t = 0.0128 secs
> n = 25, t = 0.0135 secs
> ```

With DomainMatrix you can see something more like the expected `O(n**3)` behaviour.

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_densematrix

oscarbenjamin committed:
wip: add fraction-free determinant

oscarbenjamin commented:
I've added a determinant calculation to DomainMatrix which is around 20x faster:
```python
from sympy import *
import time

for n in [1, 2, 3, 4, 5, 10, 20, 40, 80]:
    start = time.time()
    M1 = randMatrix(n)
    M1.det()
    delta = time.time() - start
    print('n = %2d, t = %.3g secs' % (n, delta))
```
On master:
```
n =  1, t = 0.000281 secs
n =  2, t = 0.000588 secs
n =  3, t = 0.000675 secs
n =  4, t = 0.00172 secs
n =  5, t = 0.00291 secs
n = 10, t = 0.0189 secs
n = 20, t = 0.144 secs
n = 40, t = 1.16 secs
n = 80, t = 9.53 secs
```
With the PR:
```
n =  1, t = 0.000467 secs
n =  2, t = 0.000608 secs
n =  3, t = 0.000489 secs
n =  4, t = 0.000572 secs
n =  5, t = 0.000809 secs
n = 10, t = 0.0023 secs
n = 20, t = 0.00886 secs
n = 40, t = 0.0532 secs
n = 80, t = 0.458 secs
```
Possibly the Bareiss algorithm would be faster but this is using fraction-free Gaussian elimination.

oscarbenjamin committed:
wip: fix det and pythonrational

oscarbenjamin commented:
I see now that the reason some benchmarks are slower is because the benchmarks are run without gmpy2 installed and the PR uses QQ even for integer matrices so we end up with pythonrational which is slow compared to mpq and also slow compare to Integer as would be used on master.

Making sure to use ZZ rather than always QQ should fix that.

oscarbenjamin committed:
wip: dont recompute gcd in pythonrational

oscarbenjamin committed:
wip: dont use floor division

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip: add matrix unify; allow symbols

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip; allow rational functions with rational coefficients

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip; fix doctests

oscarbenjamin committed:
fix tests

oscarbenjamin committed:
wip; fix tests

oscarbenjamin committed:
wip; change numpy tests

oscarbenjamin committed:
Only use a field when needed

oscarbenjamin committed:
Convert to field when needed...

oscarbenjamin committed:
Fix examples

oscarbenjamin committed:
Don't convert if the domains are the same

oscarbenjamin committed:
Fix domain check

oscarbenjamin committed:
use fraction-free algorithm for det

oscarbenjamin committed:
Use convert_from rather than convert for speed

oscarbenjamin committed:
Convert to domain in sympy_eqs_to_matrix

oscarbenjamin committed:
wip; add LU decomposition to DomainMatrix

oscarbenjamin committed:
wip; add LUdecomp

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
wip

oscarbenjamin committed:
Add charpoly to DomainMatrix

oscarbenjamin committed:
add charpoly

sylee957 commented:
Although it may be tempting to use DomainMatrix for polynomial-like expressions for computation,
I don't think that it is right to automatically convert to polynomial because I believe that there are some conceptual differences between polynomial `x` and function `x`, especially regarding the topology. 

I think that some people have different perspective of how `x / x` should automatically cancel to `1` or should it be undefined for `0`.
And similarly, `Matrix([[a, 0], [0, a]])` can be viewed as a non-singular matrix or a matrix that can either be singular or nonsingular depending on the value of `a`, which often appears in the problem of giving the exact condition when the equation have solutions or not.
Although I don't see that matrix was already designed to give such good quantifier elimination, but I think that this should be clarified because it can give an unique way how Expr matrix should be extended for.

oscarbenjamin committed:
wip flake8

oscarbenjamin commented:
I think that's why it would be good to be able to use `DomainMatrix` as a class separately from Expr-matrix although `DomainMatrix` can also be used to speed up `Expr` matrix. There is no issue with Integer/Rational or Gaussian integer or other concrete number types without symbols and this PR makes those much faster.

There are a few places around the codebase that really do want a domain version of `Matrix` and currently abuse normal Matrix by subclassing it and then invalidating its key internal invariant (that the elements are Expr):

1. There is `RawMatrix` which used to be used in `solve_lin_sys` for heurisch and is still used for the Smith normal form. This subclasses `Matrix` to prevent sympification and store domain elements directly in the matrix.
2. There is `NewMatrix` in holonomic which also stores domain elements directly in the matrix.
3. There is `PolyMatrix` which is used by risch and has elements that are sometimes Poly and sometimes Expr. This is also used by users who want to work with matrices of non-standard domains like finite fields but doesn't work properly for that (or for much at all).

All of those cases would be better served by a full-featured `DomainMatrix` class that can eliminate the need for the other subclasses. That would also free up general matrix to require that its elements be `Expr` (even if sometimes stored internally as domain elements). Then the normal matrix wouldn't need to keep working around the limitations imposed by being abused in subclasses.

If we want to handle things like conditionally existent solutions etc then there needs to be a fundamental shift in the way that sympy works and in the types of objects returned by APIs like `solve` etc. At the moment normal Expr arithmetic, polynomial roots, matrix calculations, solvers for algebraic/differential equations all return solutions according to the generic case. The changes here are a faster and more consistent way of doing that for matrix calculations.

There are some limitations to using the poly domains though like the fact that all poly-powers are expanded all the time so e.g. with this PR:
```julia
In [2]: Matrix([[(x+1)**10]])                                                                                                                  
Out[2]: Matrix([[x**10 + 10*x**9 + 45*x**8 + 120*x**7 + 210*x**6 + 252*x**5 + 210*x**4 + 120*x**3 + 45*x**2 + 10*x + 1]])
```
That's the most likely user-visible change here (apart from the speed).

oscarbenjamin commented:
I've added a `charpoly` method that implements the Berkowitz algorithm using domain elements and is significantly faster at least in some cases. For example the matrix from #19532 can be handled much faster:
```python
from sympy import *

a, b, c, d, e = symbols("a b c d e")

M = Matrix([
[      0,    1,       0,    0,       0,    0,    0],
[-1/a**2, -2/a,       0,    0,       0,    0,    0],
[      0,    0,       0,    1,       0,    0,    0],
[      0,    0, -1/b**2, -2/b,       0,    0,    0],
[      0,    0,       0,    0,       0,    1,    0],
[      0,    0,       0,    0, -1/c**2, -2/c,    0],
[    1/d,    0,     1/d,    0,     1/d,    0, -1/e]])
```
With the PR:
```julia
In [4]: %time ok=M.charpoly()                                                                                                                  
CPU times: user 14.4 ms, sys: 1.84 ms, total: 16.3 ms
Wall time: 14.8 ms
```
With master that's:
```
In [2]: %time ok=M.charpoly()                                                                                                                  
CPU times: user 3.58 s, sys: 18.4 ms, total: 3.6 s
Wall time: 3.62 s
```
So the PR makes that 200x faster.

In fact with all changes combined the exponential or Jordan form of the matrix can be computed in much smaller time than before.

On master:
```
In [2]: %time ok=M.exp()                                                                                                                       
CPU times: user 42.4 s, sys: 304 ms, total: 42.7 s
Wall time: 43.9 s
```
With the PR:
```
In [2]: %time ok=M.exp()                                                                                                                       
CPU times: user 920 ms, sys: 11 ms, total: 931 ms
Wall time: 941 ms
```
With the PR the bottleneck for eigenvalues is actually in computing the roots rather than computing the characteristic polynomial. The bulk of the time is spent in jordan form constructing the generalised eigenvectors and computing nullspaces. Converting back and forth between sympy and domain elements can slow things down a lot so if those things were all implemented for the domain elements then I think it could be maybe 4x faster for computing the exponential.

oscarbenjamin committed:
wip; add __neg__

oscarbenjamin committed:
wip; add __sub__

oscarbenjamin committed:
wip; move DomainMatrix to domainmatrix.py

oscarbenjamin committed:
wip; fix imports

oscarbenjamin committed:
Fix sub

oscarbenjamin committed:
Add DDM class

oscarbenjamin committed:
wip add DDM class

oscarbenjamin committed:
wip; add DDM class

oscarbenjamin committed:
wip; add DDM class

sylee957 commented:
I'd have a few comments to address at this stage now

1. I see that a lot of functions needed to be copied to DomainMatrix just because there are too many intermediate simplifications. How should we avoid these duplicates? At least your previous suggestion to pull out all the matrix methods doesn't look like preventing this issue.

2. I'd prefer `MatrixRingElement` than `DomainMatrix` which is more natural.
And instead of introducing only a single matrix class, I think that the ring of matrix integrated into domain computation, such that we can have `MatrixRing` to be used in other ground domains as well. But I'm not clear about the way how to introduce such noncommutative algebra or is it an intended scope about.

oscarbenjamin committed:
add rref to DDM

oscarbenjamin commented:
This is still a work in progress but yes I am aiming to come up with something that would resemble a domain element and that could be used to define a `MatrixRing` later. I've now added a DDM (dense domain matrix) class which is a list subclass. The immutable version would subclass tuple and be hashable etc.

There are functions for working with DDM like `ddm_imatmul` etc that can work with arbitrary list of list type containers so that the core algorithms can be applied to other objects without any need to subclass DDM.

The reason for copying (or rather rewriting) these routines is partly to eliminate unnecessary dependence on the implementation of the existing matrix classes and partly to make them faster. The other factor is that we can make decisions about the algorithm depending on the domain e.g. whether to use fraction free elimination when computing the determinant:
https://github.com/sympy/sympy/blob/167016351370dd418a7f423ca93e3fd974079005/sympy/polys/domainmatrix.py#L387-L393
For example I think we can make some calculations involving rref a lot faster for rings by using fraction-free elimination to reach echelon form.

oscarbenjamin committed:
wip; add det to DDM

oscarbenjamin committed:
rename from_list_sympy_2

oscarbenjamin committed:
wip; add inv to DDM

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_densematrix

oscarbenjamin committed:
wip; add lu to DDM

oscarbenjamin committed:
wip; add lu_solve to DDM

oscarbenjamin committed:
wip; add charpoly to ddm

oscarbenjamin committed:
wip; add tests for DDM

oscarbenjamin committed:
wip; add tests for ddm lu

oscarbenjamin committed:
wip; add tests for lu_solve and charpoly

oscarbenjamin committed:
wip; add tests for DomainMatrix

oscarbenjamin committed:
wip; adding tests for DomainMatrix

oscarbenjamin committed:
wip; use add rather than __add__ in DomainMatrix

oscarbenjamin committed:
wip; fix ddm_berk base case

sylee957 commented on doc/src/modules/matrices/matrices.rst:
```diff
@@ -331,24 +331,27 @@ applying ``gcd`` to the data of the matrix:
     [ 2         2      ]
     [x *y - y  x *y - y]
+    >>> gcd(tuple(_))
+    1
```
This should not be an intended result because this tutorial is actually about how to use GCDs to simplify the terms.

oscarbenjamin replied:
To be clear I am still just using this PR as a way of experimenting with DomainMatrix. I won't ask for this to be merged directly. Rather when I am done I will take out the domain matrix parts and put them in a separate PR. Ordinary matrices would (at least initially) be left alone.

sylee957 commented on sympy/polys/domainmatrix.py:
```diff
@@ -0,0 +1,653 @@
+    return qout
+
+
+class DomainMatrix:
```
I'm not sure about the purpose of doubly wrapping the classes here, but I already saw weird designs like PolynomialRing wrapping PolyRing, so there can be a reason or not.

oscarbenjamin replied:
This is still incomplete but the purpose is that DomainMatrix should function more or less like Poly and should handle unification logic for the lower-level representations. It should be possible to have a DomainMatrix that transparently uses a lower-level dense or sparse representation for a matrix with elements in any domain and those should interact nicely together with all conversions managed at the DomainMatrix level. The lower-level classes are more like the representative elements of a ring/field. You could make a domain to represent those as something like `MatrixRing(QQ, 2)` (2x2 matrices over the rationals).

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_densematrix

oscarbenjamin committed:
Disable test...

oscarbenjamin committed:
Merge remote-tracking branch 'upstream/master' into pr_densematrix

oscarbenjamin committed:
Fix float comparison in stats test

czgdp1807 commented:
@oscarbenjamin Is the purpose of this PR fulfilled? It it needed anymore?

oscarbenjamin commented:
This ended up being implemented in a more complete way in #21626
